src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,558 @@
+            // context to analyze unused suppressions using reported compiler and analyzer diagnostics.
+            // Instead, the analyzer defines a special 'AnalyzeAsync' method that should be invoked
+            // by the host with CompilationWithAnalyzers input to compute unused suppression diagnostics.
+        }
+
+        public async Task AnalyzeAsync(
+            SemanticModel semanticModel,
+            TextSpan? span,
+            CompilationWithAnalyzers compilationWithAnalyzers,
+            Func<DiagnosticAnalyzer, ImmutableArray<DiagnosticDescriptor>> getSupportedDiagnostics,
+            Func<DiagnosticAnalyzer, bool> getIsCompilationEndAnalyzer,
+            Action<Diagnostic> reportDiagnostic,
+            CancellationToken cancellationToken)
+        {
+            // We need compilation with suppressed diagnostics for this feature.
+            if (!compilationWithAnalyzers.Compilation.Options.ReportSuppressedDiagnostics)
+            {
+                return;
+            }
+
+            var tree = semanticModel.SyntaxTree;
+
+            // Bail out if analyzer is suppressed on this file or project.
+            // NOTE: Normally, we would not require this check in the analyzer as the analyzer driver has this optimization.
+            // However, this is a special analyzer that is directly invoked by the analysis host (IDE), so we do this check here.
+            if (tree.DiagnosticOptions.TryGetValue(IDEDiagnosticIds.RemoveUnnecessarySuppressionDiagnosticId, out var severity) ||
+                compilationWithAnalyzers.Compilation.Options.SpecificDiagnosticOptions.TryGetValue(IDEDiagnosticIds.RemoveUnnecessarySuppressionDiagnosticId, out severity))
+            {
+                if (severity == ReportDiagnostic.Suppress)
+                {
+                    return;
+                }
+            }
+
+            // Bail out if analyzer has been turned off through options.
+            var userExclusions = GetUserExclusions(tree, compilationWithAnalyzers, cancellationToken, out var excludeAll);
+            if (excludeAll)
+            {
+                return;
+            }
+
+            var root = tree.GetRoot(cancellationToken);
+
+            // Bail out if there are no pragma directives or tree has syntax errors.
+            if (!root.ContainsDirectives || root.GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error))
+            {
+                return;
+            }
+
+            // Process pragma directives in the tree.
+            // The core algorithm is as follows:
+            //  1. Iterate through all the active pragmas in the source file and identify the pragmas
+            //     with diagnostics IDs for which we support unnecesary pragma analysis.
+            //  2. Build the following data structures during this loop:
+            //      a. A map from diagnostic ID to list of pragmas for the ID. This map tracks supported diagnostic IDs for this tree's pragmas.
```

Comment:
> This map tracks supported diagnostic IDs for this tree's pragmas.

the latter sentence is a bit unclear to me.

Reply:
Given that diagnostic IDs are not supported by this analyzer, the map contains applicable/supported diagnostic IDs for which there are pragmas in the file.